### ISSUE TITLE
Edge-case in parsing of integer expressions with unary minus

### ISSUE THREAD

**Issue by rec0de** [abstools/abstools#283]
Hi everyone!
I've been working on automatically generating random ABS programs for a related project and I noticed that some of the generated programs that *should* compile do not.
Narrowing down the error, I eventually got to this minimal example, which I believe shows a bug in the ABS parser:

```
module MockABS;

{
	Int i =  -(0 - 1) + 1;
}
```
Replacing any of the three integer literals with an integer identifier makes the problem disappear, as does removing any part of the expression. The bug occurs only if the first literal is lower in value than the second (i.e. the result of the subtraction would be negative).

Judging from the following stack trace, the bug seems to be related to some rewrite rule for constant expressions causing a misplaced minus in a value that is parsed as an integer.

```
java.lang.NumberFormatException: Illegal embedded sign character
	at java.math.BigInteger.<init>(BigInteger.java:428)
	at java.math.BigInteger.<init>(BigInteger.java:606)
	at org.abs_models.frontend.ast.AddAddExp.rewriteRule0(AddAddExp.java:408)
	at org.abs_models.frontend.ast.AddAddExp.rewriteTo(AddAddExp.java:394)
	at org.abs_models.frontend.ast.ASTNode.getChild(ASTNode.java:768)
	at org.abs_models.frontend.ast.ASTNode.collect_contributors_Stmt_getRewritableAACs(ASTNode.java:1188)
	at org.abs_models.frontend.ast.ASTNode.collect_contributors_Stmt_getRewritableAACs(ASTNode.java:1188)
	at org.abs_models.frontend.ast.ASTNode.collect_contributors_Stmt_getRewritableAACs(ASTNode.java:1188)
	at org.abs_models.frontend.ast.Stmt.survey_Stmt_getRewritableAACs(Stmt.java:396)
	at org.abs_models.frontend.ast.Stmt.getRewritableAACs_compute(Stmt.java:541)
	at org.abs_models.frontend.ast.Stmt.getRewritableAACs(Stmt.java:519)
	at org.abs_models.frontend.ast.Stmt.hasRewritableAwaitAsyncCall(Stmt.java:143)
	at org.abs_models.frontend.ast.Block.blockHasRewritableAwaitAsyncCalls(Block.java:223)
	at org.abs_models.frontend.ast.Block.rewriteTo(Block.java:977)
	at org.abs_models.frontend.ast.ASTNode.getChild(ASTNode.java:768)
	at org.abs_models.frontend.ast.MethodImpl.getBlock(MethodImpl.java:451)
	at org.abs_models.frontend.ast.MethodImpl.refined_TypeChecker_MethodImpl_typeCheck(MethodImpl.java:142)
	at org.abs_models.frontend.ast.MethodImpl.typeCheck(MethodImpl.java:489)
	at org.abs_models.frontend.ast.ClassDecl.refined_TypeChecker_ClassDecl_typeCheck(ClassDecl.java:274)
	at org.abs_models.frontend.ast.ClassDecl.typeCheck(ClassDecl.java:1580)
	at org.abs_models.frontend.ast.ModuleDecl.typeCheck(ModuleDecl.java:112)
	at org.abs_models.frontend.ast.CompilationUnit.refined_TypeChecker_CompilationUnit_typeCheck(CompilationUnit.java:170)
	at org.abs_models.frontend.ast.CompilationUnit.typeCheck(CompilationUnit.java:1080)
	at org.abs_models.frontend.ast.Model.refined_TypeChecker_Model_typeCheck(Model.java:677)
	at org.abs_models.frontend.ast.Model.typeCheck(Model.java:1333)
	at org.abs_models.frontend.ast.Model.typeCheck_compute(Model.java:2759)
	at org.abs_models.frontend.ast.Model.typeCheck(Model.java:2746)
	at org.abs_models.frontend.parser.Main.typeCheckModel(Main.java:351)
	at org.abs_models.frontend.parser.Main.analyzeFlattenAndRewriteModel(Main.java:263)
	at org.abs_models.frontend.parser.Main.parse(Main.java:171)

```

**Comment by rudi** [comment]
Very cool! I’ve been idly thinking about fuzzing the abs toolchain; no doubt this would find some bugs.

Thanks for the report - I’ll have a look soon.

**Comment by DerDrodt** [comment]
The issue is the rewrite of `MinusExp`, Specifically the first rule:
```
when (getOperand().isConstant()
    && getOperand().getType().isIntType())
to IntLiteral new IntLiteral("-" + ((IntLiteral)getOperand()).getContent());
```
When `-(0-1)` gets rewritten, we first get `-(-1)`. This is constant and of type int, so we use the first rule and rewrite it to `--1` because we just add a "-" prefix. This is not accepted by `BigInteger` when we rewrite the outer `AddAddExp`.

This should be fixed by #284

**Comment by rudi** [comment]
Fixed in #284

**Comment by rec0de** [comment]
Here's the code generator script for reference:
https://github.com/rec0de/crowbar-investigator-experiments/blob/master/mockABS.rb
